### ISSUE TITLE
Corrupt deflate stream with GMCP

### ISSUE THREAD

**Issue by Xandaros** [LiquidityC/Blightmud#73]
After enabling GMCP with something like this:
```lua
blight:on_gmcp_ready(function()
    blight:register_gmcp("Room")
end)
```
The connection to the MUD gets closed after a few seconds, presumably when the GMCP messages arrive. In the Log, this can be found:
```
[00:00:00.000] (7f1ea2d99ac0) INFO   Starting application
[00:00:00.046] (7f1ea2d99ac0) INFO   Loading script: /home/xandaros/.config/blightmud/credentials.lua
[00:00:00.046] (7f1ea2d99ac0) INFO   Loading script: /home/xandaros/.config/blightmud/achaea.lua
[00:00:06.659] (7f1ea2212700) ERROR  Error: corrupt deflate stream
[00:03:11.787] (7f1ea2d99ac0) INFO   Shutting down
```

This is on Achaea (achaea.com:23)

**Note to dev (by LiquidityC)**
This is caused by compressed data arriving in the stream before the mccp2 setting had been toggled and decompression started. Solution is to never read beyond the `IAC SB MCCP2 IAC SE` from the tcp stream. Not an easy task.

**Comment by LiquidityC** [comment]
This one has me baffled. The IRE muds are the ones I perform most of my testing on. They are also the ones I play with the latest version of blightmud and I'm not getting any problems whilst using a lot of GMCP stuff. Including `Room`.

Are you running version 5.0? Any other middleware between you and the mud? Do you have a `report-*.toml` file in your `/tmp` dir?

**Comment by Xandaros** [comment]
I am indeed running 5.0. Installed today, from AUR.
There were a couple report files, but after deleting them and causing the problem to reoccur, no new ones were created, so I assume they were all from previous problems, when I had errors in my Lua script.
(It actually crashed the entire thing. Might not be the best, either)

Nothing else running between blightmud and the server. It is connecting directly to achaea.com:23.
My entire config file, excluding stuff I commented out:
```lua
blight:on_connect(function(host, port)
    blight:load("/home/xandaros/.config/blightmud/credentials.lua")
    if credentials[host] then
        blight:send(credentials[host].username)
        blight:send(credentials[host].password)
    end
    credentials = nil
end)

blight:on_gmcp_ready(function()
    blight:register_gmcp("Room")
end)
```

**Comment by Xandaros** [comment]
Actually, if I remove the autologin, it works fine.
I also tried adding a print in `on_gmcp_ready` and I don't ever see it. If I don't have the hook, it works fine, though, so that is confusing me quite a lot.

Is there a way to get the relevant session within the `on_gmcp_ready` hook, so I can send a `Char.Login` instead? I do need the host to retrieve the relevant credentials, although I suppose I can just hardcode it for testing purposes.

**Comment by LiquidityC** [comment]
Crashing on lua script errors isn't great. Didn't think that happened. If you can recall the errors then please post them.

As a workaround here. I'm guessing you have pretty low ping to Achaea (perhaps not relevant). But I think the problem here stems from a slightly to fast login cycle. Try disabling the credentials part and logging in manually. That will likely stop the "stream corruption". You can test with a prompt trigger also (login can't happen as fast as you're doing it is the core problem). The problem is likely that the inputstream is flooded with compressed data before blightmud has switched to decompression mode on your side.

Naturally this needs to be fixed but it'll be tricky and won't be a quick patch. So for the time being I hope my guess at your problem will help you get around it.

----
You beat me to my reply. GMCP Room data is pretty bulky. It probably impacts the tcp_stream flooding during MCCP2 initiation.

**Comment by Xandaros** [comment]
Creating an alias which, when triggered, sends a `Char.Login` GMCP message also causes the client to immediately crash because of a supposedly corrupt deflate stream.

Before it does, it seems to trigger the `on_gmcp_ready` hook a second time.

**Comment by LiquidityC** [comment]
Ok. Good information to have. I’ve got enough to debug this properly now and I was able to recreate the problem locally 😊

**Comment by LiquidityC** [comment]
I've debugged this one on my end now. The problem is what I feared. We get a bunch of uncompressed data, in the middle of that we get the `IAC SB MCCP2 IAC SE` command, after that byte sequence the rest of the data chunk is compressed. The way we do MCCP2 activation doesn't perform well with this. My initial idea is to "push back" the remainder of the bytes after the compress activation onto the incoming data (can be done through chaining).

This would allow the decompression to start as intended no matter how we receive the data. I can't se any way to control the tcp_stream reading to not read past the compress2 initiation bytes so with the current functionality I think this is the only choice. I'll keep fiddling with it.
